### PR TITLE
Added new endpoint, fix and added tests.

### PR DESCRIPTION
**Pull request by SanaviaNicolas**
### WORKSPACES
- Added new endpoint with its methods(See official bitbucket API [documentation](https://developer.atlassian.com/bitbucket/api/2/reference/resource/workspaces)):
   - list();
   - all();
   - show();
- Added tests
- Added Workspaces/Repositories with its method: show()

### PROJECTS
- Added all() method
- Fix create() and update() methods
- Fix buildProjectPath() method, use workspaces

### REPOSITORIES
- Added workspaces() method
- Added Repositories/Workspaces class with its methods:
   - createRepository();
   - showRepository();
   - buildRepositoriesPath();
- Added all() method on Team/Repositories

### TESTS
- Added mockup API features in order to test all endpoints

### CODE REVIEW

**Comment by GrahamCampbell** [comment]
This PR seems to still be based off of 2.1 and not 3.0.

File: src/Api/Repositories/Workspaces.php [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+     * @param string $repo
+     * @param $scm
+     * @param $project_key
+     * @param $is_private
+     *
+     * @return array
+     *
+     * @throws \Http\Client\Exception
+     */
+    public function createRepository(string $repo, $scm, $project_key, $is_private)
```
Comment (GrahamCampbell) [inline src/Api/Repositories/Workspaces.php]: Please use a params array like the rest of the code does.

File: src/Api/Repositories/Workspaces/AbstractWorkspacesApi.php [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+     * The repo.
+     *
+     * @var string
+     */
+    protected $repo;
+
+    /**
+     * Create a new api instance.
+     *
+     * @param \Http\Client\Common\HttpMethodsClient $client
```
Comment (GrahamCampbell) [inline src/Api/Repositories/Workspaces/AbstractWorkspacesApi.php]: this needs to be the interface, like the other 3.0 code does

File: src/Api/Teams/Projects.php [hunk 3]
```diff
@@ -23,7 +25,7 @@ class Projects extends AbstractTeamsApi
     /**
      * @param array $params
      *
-     * @throws \Http\Client\Exception
```
Comment (GrahamCampbell) [inline src/Api/Teams/Projects.php]: Please leave this alone

File: src/Api/Teams/Projects.php [hunk 4]
```diff
@@ -35,14 +37,49 @@ public function list(array $params = [])
...
+     * @param string $key
+     * @param string $description
+     * @param string $links
+     * @param bool $is_private
      *
      * @return array
+     * @throws Exception
      */
-    public function create(array $params = [])
+    public function create(string $name, string $key, string $description, string $links, bool $is_private)
```
Comment (GrahamCampbell) [inline src/Api/Teams/Projects.php]: Please revert this change.

File: tests/responses/BaseResponse.php [hunk 5]
```diff
@@ -0,0 +1,26 @@
+<?php
+
+
+namespace Bitbucket\Tests\responses;
+
+
+class BaseResponse
+{
+    /**
+     * @param $path
```
Comment (GrahamCampbell) [inline tests/responses/BaseResponse.php]: invalid syntax

**Review by GrahamCampbell — CHANGES_REQUESTED** [review]

File: src/Api/Teams/Projects.php (hunk 3 above)
Comment (SanaviaNicolas) [reply src/Api/Teams/Projects.php]: I don't understand, how can I solve this?

File: tests/responses/BaseResponse.php (hunk 5 above)
Comment (SanaviaNicolas) [reply tests/responses/BaseResponse.php]: I don't understand, how can I solve this?

File: src/Api/Teams/Projects.php (hunk 3 above)
Comment (GrahamCampbell) [reply src/Api/Teams/Projects.php]: Change it back to what it was before?

File: tests/responses/BaseResponse.php (hunk 5 above)
Comment (GrahamCampbell) [reply tests/responses/BaseResponse.php]: You are missing the type.

File: composer.json [hunk 6]
```diff
@@ -18,7 +18,8 @@
         "php-http/multipart-stream-builder": "^1.1",
         "psr/cache": "^1.0",
         "psr/http-client-implementation": "^1.0",
-        "psr/http-message": "^1.0"
+        "psr/http-message": "^1.0",
+        "php-http/mock-client": "^1.3"
```
Comment (GrahamCampbell) [inline composer.json]: Why is this in the require block and not `require-dev`?

File: composer.json [hunk 7]
```diff
@@ -38,6 +39,11 @@
     "config": {
         "preferred-install": "dist"
     },
+    "extra": {
```
Comment (GrahamCampbell) [inline composer.json]: Please remove this.

File: src/Api/Repositories.php [hunk 8]
```diff
@@ -46,6 +47,16 @@ public function users(string $username)
...
     }
 
+    /**
+     * @param string $username
+     *
+     * @return \Bitbucket\Api\Repositories\Workspaces
+     */
+    public function workspaces(string $username)
+    {
+        return new RepositoriesWorkspaces($this->getHttpClient(), $username);
```
Comment (GrahamCampbell) [inline src/Api/Repositories.php:57]: I don't understand how this is any different to the `users` API?

**Review by GrahamCampbell — CHANGES_REQUESTED** [review]

File: composer.json (hunk 6 above)
Comment (Giacomo92) [reply composer.json:22]: I have added these dependencies in order to make tests. I have mock the client instance so every request made in test section does not call Bitbucket directly but returns a static response saved in a json file

**Comment by GrahamCampbell** [comment]
> Added Workspaces/Repositories with its method: show()

**Comment by GrahamCampbell** [comment]
> Added Workspaces/Repositories with its method: show()

All this stuff already exists, under the name "users". Please don't duplicate it. I will likely rename everything in 3.0 to match Bitbucket's API changes.

File: src/Api/Teams/Projects.php [hunk 9]
```diff
@@ -34,10 +36,29 @@ public function list(array $params = [])
         return $this->get($path, $params);
     }
 
+    /**
+     * @throws Exception
+     *
+     * @return array
+     */
+    public function all()
```
Comment (GrahamCampbell) [inline src/Api/Teams/Projects.php]: Please delete this method. Fetching all is already supported by the `ResultPager` class.

File: src/Api/Teams/Projects.php [hunk 10]
```diff
@@ -52,7 +73,7 @@ public function create(array $params = [])
      * @param string $project
      * @param array  $params
      *
-     * @throws \Http\Client\Exception
```
Comment (GrahamCampbell) [inline src/Api/Teams/Projects.php]: Please revert this change and all related changes.

File: src/Api/Teams/Projects.php [hunk 11]
```diff
@@ -67,11 +88,11 @@ public function show(string $project, array $params = [])
...
+     * @throws Exception
      *
      * @return array
      */
-    public function update(string $project, array $params = [])
```
Comment (GrahamCampbell) [inline src/Api/Teams/Projects.php]: Please revert that.

File: src/Api/Teams/Projects.php [hunk 12]
```diff
@@ -104,6 +125,6 @@ public function remove(string $project, array $params = [])
      */
     protected function buildProjectsPath(string ...$parts)
     {
-        return static::buildPath('teams', $this->username, 'projects', ...$parts);
+        return static::buildPath('workspaces', $this->username, 'projects', ...$parts);
```
Comment (GrahamCampbell) [inline src/Api/Teams/Projects.php]: Please don't change this.

File: src/Api/Teams/Repositories.php [hunk 13]
```diff
@@ -34,6 +34,25 @@ public function list(array $params = [])
         return $this->get($path, $params);
     }
 
+    /**
+     * @throws \Http\Client\Exception
+     *
+     * @return array
+     */
+    public function all()
```
Comment (GrahamCampbell) [inline src/Api/Teams/Repositories.php]: Please delete this.

File: src/Api/Workspaces.php [hunk 14]
```diff
@@ -0,0 +1,74 @@
...
+     * @return array
+     */
+    public function list(array $params = [])
+    {
+        $path = $this->buildWorkspacePath();
+
+        return $this->get($path, $params);
+    }
+
+    public function all()
```
Comment (GrahamCampbell) [inline src/Api/Workspaces.php]: As before, we don't need these all methods. We already have this implemented in a way that respects separation of concerns.

**Review by GrahamCampbell — CHANGES_REQUESTED** [review]

**Comment by GrahamCampbell** [comment]
I've added some examples to the README now, including how to use the paginator, which is why those `all` methods aren't needed: https://github.com/BitbucketAPI/Client/tree/6b8785635eeae87de7edd682968a0e305e5ab406#example-3.

**Comment by Giacomo92** [comment]
@GrahamCampbell thank you for all the feedbacks. We will fix them soon. What do you think about test system? Can we add more test (like we have already did in this PR)?

**Comment by GrahamCampbell** [comment]
The tests are great. In fact, I might take over the `src` changes, since there are quite a lot of other changes not covered by this PR that I'd like to make. After I'm finished, if you rebase this PR, we should be just left with the tests.

**Comment by GrahamCampbell** [comment]
Ok, I've finished rigorously replacing teams/users stuff with workspaces: https://github.com/BitbucketAPI/Client/compare/2.1...3.0. I've also removed any deprecated endpoints. I just have left to go through the repositories endpoints to see if there's anything new they've added we don't yet support.

If you could rebase this PR, just providing the tests, that would be great. I'll then given them a review and we hopefully can merge this PR. :)

File: src/Api/Repositories/Workspaces/Repositories.php [hunk 15]
```diff
@@ -0,0 +1,51 @@
...
+ */
+
+namespace Bitbucket\Api\Repositories\Workspaces;
+
+/**
+ * The users api class.
+ *
+ * @author Graham Campbell <graham@alt-three.com>
+ */
+class Repositories extends AbstractWorkspacesApi
```
Comment (GrahamCampbell) [inline src/Api/Repositories/Workspaces/Repositories.php:21]: Please delete this. There is no such API.

File: src/Api/Workspaces.php [hunk 16]
```diff
@@ -60,6 +60,20 @@ public function show(array $params = [])
...
     }
 
+    /**
+     * @param array $params
+     *
+     * @throws \Http\Client\Exception
+     *
+     * @return array
+     */
+    public function list(array $params = [])
```
Comment (GrahamCampbell) [inline src/Api/Workspaces.php:70]: Please delete this. This is already implemented on the "CurrentUser" API.

File: src/Api/Workspaces/Projects.php [hunk 17]
```diff
@@ -71,7 +71,7 @@ public function show(string $project, array $params = [])
      *
      * @return array
      */
-    public function update(string $project, array $params = [])
```
Comment (GrahamCampbell) [inline src/Api/Workspaces/Projects.php:74]: Please don't change this. We have params like this out of convention. If we want to review that, we'd want to do it for all cases, not just one, and not in this PR.

File: src/Api/Workspaces/Projects.php [hunk 18]
```diff
@@ -104,6 +104,6 @@ public function remove(string $project, array $params = [])
      */
     protected function buildProjectsPath(string ...$parts)
     {
-        return static::buildPath('workspaces', $this->username, 'projects', ...$parts);
```
Comment (GrahamCampbell) [inline src/Api/Workspaces/Projects.php:107]: This typo, and a few others are already fixed on `3.0`.

**Comment by GrahamCampbell** [comment]
Thanks for the PR. I'll finish off here, and merge locally. :)

File: tests/responses/ProjectsAllResponse.php [hunk 19]
```diff
@@ -0,0 +1,26 @@
+<?php
+
+namespace Bitbucket\Tests\responses;
+
+use GuzzleHttp\Psr7\Response;
+use Http\Message\ResponseFactory;
+use Psr\Http\Message\ResponseInterface;
+
+class ProjectsAllResponse extends BaseResponse implements ResponseFactory
```
Comment (GrahamCampbell) [inline tests/responses/ProjectsAllResponse.php:9]: Wrong to implement this interface, when you then disregard the parameters.

**Comment by Giacomo92** [comment]
Hi @GrahamCampbell , I have just seen the new PR. Good job!
Do you need an hand in to increase code coverage?

**Comment by GrahamCampbell** [comment]
I don't think getting 100% code coverage with that style of test is really going to add much value. It was certainly good to have a small number of those tests though, like you added, so thanks for that.

Those sorts of tests don't catch mistakes URLs generated by the code, which is the most common error, I think.

**Comment by Giacomo92** [comment]
Can you give me an example of "mistakes URLs generated by the code"?

**Comment by GrahamCampbell** [comment]
https://github.com/BitbucketAPI/Client/issues/15

**Comment by GrahamCampbell** [comment]
The only way to catch those sorts of errors, is to actually hit the API, like in https://github.com/BitbucketAPI/Client/blob/46bf30101114dae032ed381642e91c2ba1e1a51a/tests/ClientTest.php#L35-L47. But, obviously our tests are going to be quite slow if we do this for all the endpoints. Anyway, I don't think we should. I think we have enough coverage from people using the code and letting us know.

One place where I think we could have better test coverage is the http plugins, however.
